receiver/mongodbatlasreceiver/alerts.go

```diff
@@ -289,6 +519,133 @@ func payloadToLogs(now time.Time, payload []byte) (plog.Logs, error) {
 	return logs, nil
 }
 
+// startCheckpointer is a subprocesses that will go through the cached alerts to see which
+// ones we can remove from the cache. Once finished, it will checkpoint current progress
+func (a *alertsReceiver) startCheckpointer(ctx context.Context) {
+	defer a.wg.Done()
+	// check every 30 minutes
+	t := time.NewTicker(time.Minute * 30)
+	for {
+		select {
+		case <-t.C:
+			now := time.Now()
+			a.cache.Clean(now)
+			if err := a.writeCheckpoint(ctx); err != nil {
+				a.logger.Error("error writing a checkpoint", zap.Error(err))
+			}
+		case <-a.checkpointDoneChan:
+			return
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+type cachePutOptions struct {
+	Expires bool          `mapstructure:"expires"`
+	TTL     time.Duration `mapstructure:"ttl"`
+}
+
+// alertCache wraps a sync Map so it is goroutine safe as well as
+// can have custom marshaling
+type alertCache struct {
+	data sync.Map `mapstructure:"data"`
+}
+
+func (a *alertCache) UnmarshalJSON(data []byte) error {
+	var candidate map[string]any
+	if err := json.Unmarshal(data, &candidate); err != nil {
+		return err
+	}
+	for k, v := range candidate {
+		a.data.Store(k, v)
+	}
+	return nil
+}
+
+func (a *alertCache) MarshalJSON() ([]byte, error) {
+	candidate := make(map[string]any)
+	a.data.Range(func(key, value any) bool {
+		k, _ := key.(string)
+		candidate[k] = value
+		return true
+	})
```

Comment:
Not sure this is actually correct; sync.Map is hard to wrap my head around:

See: https://pkg.go.dev/sync#Map.Range
```
Range does not necessarily correspond to any consistent snapshot of the Map's contents: no key will be visited more than once, but if the value for any key is stored or deleted concurrently (including by f), Range may reflect any mapping for that key from any point during the Range call
```

Which could maybe result in alerts firing twice instead of respecting the TTL? Not sure. It might be better to just use a mux + primitive map in this case.

Reply:
Switched to map plus normal mutex 👍